Added detection of Escaped XSS encodeURIComponent

devplayer55221 opened the pull request:
This is to enhance the detection capability of the active scanner for XSS.
[Escaped XSS - /js/encodeURIComponent](https://www.zaproxy.org/docs/scans/firingrange/#section-escape)

- Added a payload `alert(1)` in directAttack function
- Found that the function was not getting called as different context was being detected.
- So, added the code to call directAttack function in the closeTag context function.

devplayer55221 commented:
I have read the CLA Document and I hereby sign the CLA

kingthorin commented:
Should have unit tests for the new situation/coverage, as well as a chaneglog note.

kingthorin commented:
`  Run './gradlew :addOns:ascanrules:spotlessApply' to fix these violations.

psiinon commented:
![Logo](https://cdn.ast.checkmarx.net/integrations/logo/Checkmarx.png)
**Checkmarx One – Scan Summary & Details** – [e5c93134\-e8f7\-4351\-9cd6\-e1d8501bc4a6](https://eu.ast.checkmarx.net/projects/0da811fa-6373-42b1-bbb6-d0026285b136/scans?id=e5c93134-e8f7-4351-9cd6-e1d8501bc4a6&branch=devplayer55221%2FescapedXSS%2FencodeURIComponent)
### Great job\! No new security vulnerabilities introduced in this pull request

thc202 commented:
> Added a payload `alert(1)` in directAttack function

The payload is being added everywhere, see usages of `GENERIC_SCRIPT_ALERT_LIST`.

devplayer55221 commented:
> Should have unit tests for the new situation/coverage, as well as a chaneglog note.
> 
>     * https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java
> 
>     * https://github.com/zaproxy/zap-extensions/blob/main/addOns/ascanrules/CHANGELOG.md

I need to add code for the new unit test? I am not familiar with this.

devplayer55221 commented:
> > Added a payload `alert(1)` in directAttack function
> 
> The payload is being added everywhere, see usages of `GENERIC_SCRIPT_ALERT_LIST`.

Where can I see the usage of `GENERIC_SCRIPT_ALERT_LIST`. Also can I create a new context for this type of case?

kingthorin commented:
> I run this command. There were no errors. Build was successful.

This applies the code formatting requirements. You can commit and push again.

> I need to add code for the new unit test? I am not familiar with this

Yup there are lots of examples there already.

> Where can I see the usage of GENERIC_SCRIPT_ALERT_LIST. Also can I create a new context for this type of case?

In the rule you're modifying.
Do you mean you also want to change the HTML context analyzer?

devplayer55221 commented:
> > I run this command. There were no errors. Build was successful.
> 
> This applies the code formatting requirements. You can commit and push again.
> 
> > I need to add code for the new unit test? I am not familiar with this
> 
> Yup there are lots of examples there already.
> 
> > Where can I see the usage of GENERIC_SCRIPT_ALERT_LIST. Also can I create a new context for this type of case?
> 
> In the rule you're modifying. Do you mean you also want to change the HTML context analyzer?

Okay, I will try to write a unit test.
I have changed the code where I added the payload to GENERIC_SCRIPT_ALERT_LIST. Now I am using the payload directly in a new context inside the directAttack method.
Also added the modification in Changelog file.

kingthorin commented:
I likely won't get a chance to check again until the morning, the change note should be in the Unreleased section (the version and date are set at release).

thc202 commented:
Use Java 17 as mentioned in https://www.zaproxy.org/docs/developer/quick-start-build/#preparation

devplayer55221 commented:
> Use Java 17 as mentioned in https://www.zaproxy.org/docs/developer/quick-start-build/#preparation

Java 17.0.12 should work right?
There is error regarding `:addOns:selenium:generateZapAddOnManifest`

thc202 commented:
Yes, what command are you using?

thc202 commented:
Depends on the tests, you need to check the assertions and see if the tests need to be updated (e.g. other attack being used) or the new code changes corrected (e.g. introduces FPs).

kingthorin commented:
This is how you'd test just the one class from the CLI:
```
./gradlew :aO:ascanrules:test --tests org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest
```

I can confirm on `main` it passes while on your branch (`devplayer55221:escapedXSS/encodeURIComponent`) 5 fail.

devplayer55221 commented:
> This is how you'd test just the one class from the CLI:
> 
> ```
> ./gradlew :aO:ascanrules:test --tests org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest
> ```
> 
> I can confirm on `main` it passes while on your branch (`devplayer55221:escapedXSS/encodeURIComponent`) 5 fail.
> 
> Edit: Two of them seem to be an unexpected attack value. Three of them seem to be false positives (alert raised where none is expected).

Got it, Thanks!

devplayer55221 commented:
I have made a new commit. I added my code to the scriptAttack function instead of directAttack. It woks correctly now, and unit tests also passes.

devplayer55221 commented:
Added a new commit with unit test added for XSS detection in eval in script.
Done `spotlessApply`.

thc202 commented:
The `spotlessApply` is still needed.

kingthorin commented on addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRule.java:
```diff
@@ -386,8 +389,10 @@ private void raiseAlert(int confidence, String param, HtmlContext ctx, String ot
     }
 
     private boolean performDirectAttack(HttpMessage msg, String param) {
+
```
revert

kingthorin commented on addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java:
```diff
@@ -2670,6 +2670,48 @@ protected Response serve(IHTTPSession session) {
+                                            + q
+                                            + "'));\n"
+                                            + "    </script>\n"
+                                            + "  \n"
+                                            + "\n"
```
remove

kingthorin commented on addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java:
```diff
@@ -2670,6 +2670,48 @@ protected Response serve(IHTTPSession session) {
+                        String response;
+                        if (q != null) {
+                            // Make the eye catchers fail
+                            response =
+                                    "<html><head></head><body>\n"
+                                            + "    <script>\n"
+                                            + "      eval(encodeURIComponent('"
+                                            + q
+                                            + "'));\n"
+                                            + "    </script>\n"
+                                            + "  \n"
+                                            + "\n"
+                                            + "</body></html>";
```
```suggestion
                            response = """
                            <html><head></head><body>
                                <script>
                                   eval(encodeURIComponent('%s'));
                                </script></body></html>""".formatted(q);
```

The formatting here might not be 100%, I did it by hand. But the point is it could use a text block instead of concatenation and be formatted to include the `q` value at `%s`.

devplayer55221 replied:
> The formatting here might not be 100%, I did it by hand. But the point is it could use a text block instead of concatenation and be formatted to include the `q` value at `%s`.

I made the change as suggested, and also reverted the extra newlines.

devplayer55221 commented:
> This one was not yet addressed [#6686 (comment)](https://github.com/zaproxy/zap-extensions/pull/6686#issuecomment-3223061920)

Okay, got it. I have done the spotlessApply, earlier something was missed I guess.

kingthorin commented:
```
CrossSiteScriptingScanRuleUnitTest > shouldNotAlertXssInFilteredJsString() FAILED
    java.lang.AssertionError: 
    Expected: a collection with size <0>
         but: collection size was <1>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest.shouldNotAlertXssInFilteredJsString(CrossSiteScriptingScanRuleUnitTest.java:2141)
CrossSiteScriptingScanRuleUnitTest > shouldNotReportXssInsideInputAndScriptWithGoodFiltering() FAILED
    java.lang.AssertionError: 
    Expected: <0>
         but: was <1>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest.shouldNotReportXssInsideInputAndScriptWithGoodFiltering(CrossSiteScriptingScanRuleUnitTest.java:1693)
CrossSiteScriptingScanRuleUnitTest > shouldNotAlertXssInJsStringWithEncoding() FAILED
    java.lang.AssertionError: 
    Expected: <0>
         but: was <1>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest.shouldNotAlertXssInJsStringWithEncoding(CrossSiteScriptingScanRuleUnitTest.java:2111)
```

devplayer55221 commented:
> ```
> CrossSiteScriptingScanRuleUnitTest > shouldNotAlertXssInFilteredJsString() FAILED
>     java.lang.AssertionError: 
>     Expected: a collection with size <0>
>          but: collection size was <1>
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
>         at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest.shouldNotAlertXssInFilteredJsString(CrossSiteScriptingScanRuleUnitTest.java:2141)
> CrossSiteScriptingScanRuleUnitTest > shouldNotReportXssInsideInputAndScriptWithGoodFiltering() FAILED
>     java.lang.AssertionError: 
>     Expected: <0>
>          but: was <1>
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
>         at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest.shouldNotReportXssInsideInputAndScriptWithGoodFiltering(CrossSiteScriptingScanRuleUnitTest.java:1693)
> CrossSiteScriptingScanRuleUnitTest > shouldNotAlertXssInJsStringWithEncoding() FAILED
>     java.lang.AssertionError: 
>     Expected: <0>
>          but: was <1>
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
>         at org.zaproxy.zap.extension.ascanrules.CrossSiteScriptingScanRuleUnitTest.shouldNotAlertXssInJsStringWithEncoding(CrossSiteScriptingScanRuleUnitTest.java:2111)
> ```

To resolve this, I should focus on changing my scan rule, or changing these unit tests?

kingthorin commented:
Well it depends why they're now failing. Given that no alert is raised at all it "seems" like it is something in the new changes that needs to be changed further.

devplayer55221 commented:
> Have you re-built the add-on and tested it against Firing Range?

Yes, I have checked it. It works okay. Checked it just now again.

<img width="1905" height="445" alt="image" src="https://github.com/user-attachments/assets/6fa6915e-2601-4901-b750-b5b62be5f386" />

devplayer55221 commented:
> Sorry my earlier message was wrong, it should have said: "Given that alerts are being raised where they previously weren't it "seems" like it is something in the new changes that needs to be changed/handled further."
> 
> I had a quick look at all three tests that are failing, it seems the new payload is always being assigned to a var which doesn't really represent an XSS, ex: `<script>var a = "alert(1)"</script>`

Okay, I understand. Thanks

devplayer55221 commented:
> Using `println` you should just see it in the terminal/console. Are you using Eclipse as your IDE?

I am using VS Code, and executing the commands through terminal.

kingthorin commented:
...just checked. It seems you do need to use `--info`, it's going to include a lot of output you weren't previously getting so make sure your print statements are really obvious or search for them in the terminal output.
<img width="751" height="313" alt="image" src="https://github.com/user-attachments/assets/1bd3fc63-9e85-495d-9c12-10a6b6d5b5d5" />

devplayer55221 commented:
Hi, I also just checked. I am able to get the `System.out.println` output with the use of --info in ./gradlew run.
However, I am stuck with how do I identify if `eval` is being used to execute the input. Any help on this?

devplayer55221 commented:
> Hmmm that might be easier said than done. You could try with regex but there's a ton of things to consider: multiline, minified, deeper set of calls... It might require something more like ([ANTLR](https://github.com/zaproxy/zap-extensions/pull/3838) which is currently delayed due to other changes/priorities) or graal parser.

Okay, I understand.
But I was thinking that are those unit tests that were failing, totally correct?
They are filtering the input value, but my payload is `alert(1)`, which should give XSS alert in the `eval` case.

kingthorin commented:
If we knew that the variable the payload is assigned to is being executed in an eval.

kingthorin commented:
Or we'd have to know that something like `eval(a)` ends up happening.

devplayer55221 commented:
Yeah got it. I have been debugging, and I will have to think of the logic again, it doesn't seem entirely correct.
Now I know how to debug and do unit testing, I will check further.

devplayer55221 commented:
Are contexts interpreted differently when run by unit test and in the application?
For the performScriptAttack, I am getting different responses.

When I run active scanner in application: contexts2 is empty
<img width="1282" height="683" alt="image" src="https://github.com/user-attachments/assets/7cff1f03-4b18-4c77-a84e-c47c341a448b" />

When I run the ascanrules unit tests: contexts2 is not empty
<img width="1384" height="638" alt="image" src="https://github.com/user-attachments/assets/11c77453-b960-4643-a23a-9b1f51635a82" />

The unit test function is 'shouldReportXssInSriptAttackInEval'. I could not figure out why is there a difference between active scanner and unit test for getting the value of 'contexts2'.

psiinon commented:
No, contexts should be handled the same in unit tests and in the app, otherwise the unit tests would be meaningless 😁
